gradio/blocks.py

```diff
@@ -208,29 +208,46 @@ def render(self):
         self._id = Context.id
         Context.id += 1
 
-    def process_api(self, data: Dict[str, Any], username: str = None) -> Dict[str, Any]:
+    def process_api(
+        self,
+        data: Dict[str, Any],
+        username: str = None,
+        state: Optional[Dict[int, any]] = None,
+    ) -> Dict[str, Any]:
         raw_input = data["data"]
         fn_index = data["fn_index"]
+        session_hash = data["session_hash"]
         fn, preprocess, postprocess = self.fns[fn_index]
         dependency = self.dependencies[fn_index]
 
         if preprocess:
-            processed_input = [
-                self.blocks[input_id].preprocess(raw_input[i])
-                for i, input_id in enumerate(dependency["inputs"])
-            ]
+            processed_input = []
+            for i, input_id in enumerate(dependency["inputs"]):
+                block = self.blocks[input_id]
+                if getattr(block, "stateful", False):
+                    processed_input.append(state.get(input_id))
+                else:
+                    processed_input.append(block.preprocess(raw_input[i]))
             predictions = fn(*processed_input)
         else:
             predictions = fn(*raw_input)
         if len(dependency["outputs"]) == 1:
             predictions = (predictions,)
         if postprocess:
-            predictions = [
-                self.blocks[output_id].postprocess(predictions[i])
-                if predictions[i] is not None
-                else None
-                for i, output_id in enumerate(dependency["outputs"])
-            ]
+            output = []
+            for i, output_id in enumerate(dependency["outputs"]):
+                block = self.blocks[output_id]
+                if getattr(block, "stateful", False):
+                    state[output_id] = predictions[i]
+                    output.append(None)
```

Comment:
why not append it to output?

Reply:
We don't want to send the state to the frontend. Variables are not displayed in the frontend, and may be massive or binary data. We have to append to preprocess because it is needed to run the function, but variable data is never sent to the frontend.